CHANGELOG.md

```diff
@@ -1,6 +1,27 @@
 Note: This is in reverse chronological order, so newer entries are added to the top.
 
 
+Swift 5.3.1
+-----------
+
+* [#2937](https://github.com/apple/swift-package-manager/pull/2937)
+  
+  Manifest files can now have any combination of leading line terminators. All [Unicode line terminators](https://www.unicode.org/reports/tr14/) are recognised. Gone are ye olde days when a shebang-like directive must be at the start of file.
+  
+  [SR-13566](https://bugs.swift.org/browse/SR-13566) The Swift tools version specification in each manifest file now accepts any combination of _horizontal_ whitespace characters between `//` and `swift-tools-version`, if and only if the specified version > 5.3. For example, `//swift-tools-version:5.3.1` and `//		 swift-tools-version:5.4` are valid.
+  
+  * Deprecations
+    
+    `ToolsVersionLoader.split(_ bytes: ByteString) -> (versionSpecifier: String?, rest: [UInt8])` is now deprecated, and replaced by `ToolsVersionLoader.split(_ manifestContents: String) -> ManifestComponents`.
+  
+    `ToolsVersionLoader.Error.malformedToolsVersion(specifier: String, currentToolsVersion: ToolsVersion)` is not deprecated, and replaced by `ToolsVersionLoader.Error.malformedToolsVersionSpecification(_ malformation: ToolsVersionSpecificationMalformation)`.
+  
+  * Source Breakages
+    
+    SPM now throws an error if a manifest file contains invalid byte sequences such as `0x7F8F`.
+    
+    Swift tools version specifications that contain line terminators other than `U+000A` before either "swift-tool" or "tool-version" now silently falls back to using Swift 3.1 as the lowest supported version.
```

Comment:
Is this just to maintain compatibility with older versions of SwiftPM?  If so, it would be great to call out that motivation here.  If that's not the reason, then I might have misunderstood some of the discussion in the conversation for this PR.

Reply:
Could you clarify which lines you're referring to? In my email, it shows that the question is regarding the entire entry in the changelog. On this PR page, GitHub is only highlighting the description of the 2 source breakages.

Reply:
Ah, sorry.  I was referencing the "now silently falls back to using Swift 3.1" part.  I am not sure I'm clear on what the semantics are here after the latest changes, but I think it would be a poor user experience if there were new cases in which SwiftPM silently used Swift 3.1 semantics, since that can cause downstream errors that are hard to diagnose.

My basic concern (which I have perhaps expressed poorly in previous comments) is if we're making it more likely that package authors will unintentionally cause older SwiftPM versions to silently fall back to using SwiftPM 3.1 semantics rather than emit an error.  That's my only concern.

Reply:
> if we're making it more likely that package authors will unintentionally cause older SwiftPM versions to silently fall back to using SwiftPM 3.1 semantics rather than emit an error.

This is what I think as well. I think I didn't express my intent clearly previously. Sorry.

I updated the changelog to reflect what SwiftPM's behaviour was after commit 1ef5badfd623827c8176acea0cc760ddfb8eaa7e. My thought was that if I keep the changelog in pace with the actual changes, then it will make things easier if any commit needs to be reverted.

My plan is to remove the silent fallback completely, and emit an error to inform the user of any malformation in the manifest in this PR (still working on this part). I'm not sure if we are using "older versions of SwiftPM" in the same sense, but my idea is basically this:

- If a user's current swift toolchain version ≤ 5.3, then SwiftPM's behaviour is unchanged, regardless of what the lowest supported version specified in the manifest is.

- If a user's current swift toolchain version ≥ 5.3.1, then SwiftPM never falls back at all. Instead, it always emits an error to the user if there is something wrong with the manifest file.

^ This is assuming that SwiftPM's version always follows the Swift toolchain's version. If SwiftPM comes separately from the toolchain, then the watershed would be the next SwiftPM release.

Reply:
> I updated the changelog to reflect what SwiftPM's behaviour was after commit 1ef5bad. My thought was that if I keep the changelog in pace with the actual changes, then it will make things easier if any commit needs to be reverted.

That makes a lot of sense.  Thank you for being so careful with this!

Removing the fallback makes sense to me.  Falling back to 3.1 made sense when there were packages that were so old that they didn't have a tools version comment at all.  But since SwiftPM 5.0 (commit 3fedbbd), it seems that SwiftPM anyway emits `error: package at '/path/to/package' is using Swift tools version 3.1.0 which is no longer supported`.  So this isn't really the compatibility problem that I'd thought, and making that error instead point out the actual error (such as a missing tools version designation) makes a lot of sense.  Thanks for taking this on!

Reply:
I didn't know SwiftPM already dropped support for Swift < 4. Thanks for pointing me to the commit.

In the commit message of 3fedbbd69dbcea1eb6a3ee479345712c7687940c:

> We were only doing this for root packages, we also need to do this check during dependency resolution.

I'm not completely sure, but I think the change in this PR will affect all packages (root and dependencies). If this is true, then I think it is still a source break.

Reply:
The fallbacks to 3.1 are fully removed in commit b370960f12434ec523a84bb6856ea773842edb2a.

Reply:
It looks to me as if the commit I referenced extended the check for < 4 to apply to all packages, not just the root one.  So I think we're fine.